coral-spark/src/main/java/com/linkedin/coral/spark/TransportableUDFMap.java

```diff
@@ -119,20 +137,37 @@ private TransportableUDFMap() {
    * @return Optional<SparkUDFInfo>
    */
   static Optional<SparkUDFInfo> lookup(String className) {
-    return Optional.ofNullable(UDF_MAP.get(className));
+    SCALA_VERSION scalaVersion = getScalaVersion();
+    return Optional.ofNullable(UDF_MAP.get(className)).map(scalaMap -> Optional.ofNullable(scalaMap.get(scalaVersion))
+        .orElseThrow(() -> new UnsupportedUDFException(String.format(
+            "Transport UDF for class '%s' is not supported for scala %s, please contact " + "the UDF owner for upgrade",
+            className, scalaVersion.toString()))));
+  }
+
+  public static void add(String className, String sparkClassName, String artifactoryUrlSpark211,
+      String artifactoryUrlSpark212) {
+    Map scalaToTransportUdfMap = new HashMap<SCALA_VERSION, SparkUDFInfo>() {
+      {
+        put(SCALA_VERSION.SCALA_2_11, artifactoryUrlSpark211 == null ? null : new SparkUDFInfo(sparkClassName, null,
+            Collections.singletonList(URI.create(artifactoryUrlSpark211)), SparkUDFInfo.UDFTYPE.TRANSPORTABLE_UDF));
+        put(SCALA_VERSION.SCALA_2_12, artifactoryUrlSpark212 == null ? null : new SparkUDFInfo(sparkClassName, null,
+            Collections.singletonList(URI.create(artifactoryUrlSpark212)), SparkUDFInfo.UDFTYPE.TRANSPORTABLE_UDF));
+      }
+    };
+    UDF_MAP.put(className, scalaToTransportUdfMap);
   }
 
-  public static void add(String className, String sparkClassName, String artifcatoryUrl) {
+  private static SCALA_VERSION getScalaVersion() {
     try {
-      URI url = new URI(artifcatoryUrl);
-      List<URI> listOfUris = new LinkedList<>();
-      listOfUris.add(url);
-
-      // Set the function name to null here because it is determined dynamically from the enclosing SqlOperand
-      UDF_MAP.put(className,
-          new SparkUDFInfo(sparkClassName, null, listOfUris, SparkUDFInfo.UDFTYPE.TRANSPORTABLE_UDF));
-    } catch (URISyntaxException e) {
-      throw new RuntimeException(String.format("Artifactory URL is malformed %s", artifcatoryUrl), e);
+      String sparkVersion = SparkSession.active().version();
+      if (sparkVersion.matches("2.*"))
+        return SCALA_VERSION.SCALA_2_11;
+      if (sparkVersion.matches("3.*"))
+        return SCALA_VERSION.SCALA_2_12;
```

Comment:
Can the pattern matching be a little more strict?

Reply:
Like 2.4.* and 3.1.*? 

It would be too strict, if we release 3.2.* it would again require code changes. The goal of this was to just determine scala version.

Reply:
This regex matches `2` or `3` followed by anything. It does not detect whether it is the major version for example.

Reply:
added regex to match number followed by a dot followed by multiple combinations of dot and numbers